### PR TITLE
[SPARK-41767][CONNECT][PYTHON] Implement `Column.{withField, dropFields}`

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
Implement `Column.{withField, dropFields}`

### Why are the changes needed?
For API coverage

### Does this PR introduce _any_ user-facing change?
yes

### How was this patch tested?
added UT

### CODE REVIEW

**Comment by zhengruifeng** [comment]
cc @HyukjinKwon @cloud-fan  @grundprinzip

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto [hunk 1]
```diff
@@ -241,6 +242,20 @@ message Expression {
     Expression extraction = 2;
   }
 
+  // Add, replace or drop field of `StructType` into `col` expression by name.
```
Comment (grundprinzip) [inline connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:245]: ... a field ...

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto [hunk 2]
```diff
@@ -241,6 +242,20 @@ message Expression {
...
+  message UpdateFields {
+    // (Required) The struct expression
```
Comment (grundprinzip) [inline connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:247]: .

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto [hunk 3]
```diff
@@ -241,6 +242,20 @@ message Expression {
...
+    Expression struct_expression = 1;
+
+    // (Required) The field name
```
Comment (grundprinzip) [inline connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:250]: .

File: python/pyspark/sql/connect/expressions.py [hunk 4]
```diff
@@ -420,6 +420,59 @@ def __repr__(self) -> str:
...
+        self._valueExpr = valueExpr
+
+    def to_plan(self, session: "SparkConnectClient") -> proto.Expression:
+        expr = proto.Expression()
+        expr.update_fields.struct_expression.CopyFrom(self._structExpr.to_plan(session))
+        expr.update_fields.field_name = self._fieldName
+        expr.update_fields.value_expression.CopyFrom(self._valueExpr.to_plan(session))
+        return expr
+
+    def __repr__(self) -> str:
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/expressions.py:448]: Shouldn't this be handled automatically now?

File: python/pyspark/sql/connect/expressions.py [hunk 5]
```diff
@@ -420,6 +420,59 @@ def __repr__(self) -> str:
...
+        self._fieldName = fieldName
+
+    def to_plan(self, session: "SparkConnectClient") -> proto.Expression:
+        expr = proto.Expression()
+        expr.update_fields.struct_expression.CopyFrom(self._structExpr.to_plan(session))
+        expr.update_fields.field_name = self._fieldName
+        return expr
+
+    def __repr__(self) -> str:
+        return f"DropField({self._structExpr}, {self._fieldName})"
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/expressions.py:473]: Same question?

File: python/pyspark/sql/tests/connect/test_connect_column.py [hunk 6]
```diff
@@ -653,19 +676,118 @@ def test_column_accessor(self):
             sdf.select(sdf.c[0:1], sdf["c"][2:10]).toPandas(),
         )
 
-    def test_unsupported_functions(self):
```
Comment (grundprinzip) [inline python/pyspark/sql/tests/connect/test_connect_column.py:656]: Were those three last ones? Nice!!

**Review by grundprinzip — APPROVED** [review]
Minor nits

File: python/pyspark/sql/tests/connect/test_connect_column.py (hunk 6 above)
Comment (zhengruifeng) [reply python/pyspark/sql/tests/connect/test_connect_column.py:656]: `__iter__` is not supported, I just move it to another place.

PySpark did not support `__iter__` either

File: python/pyspark/sql/connect/expressions.py (hunk 4 above)
Comment (zhengruifeng) [reply python/pyspark/sql/connect/expressions.py:448]: yes, let me remove it

File: python/pyspark/sql/connect/expressions.py (hunk 5 above)
Comment (zhengruifeng) [reply python/pyspark/sql/connect/expressions.py:473]: will remove it

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto (hunk 1 above)
Comment (zhengruifeng) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:245]: will update

File: python/pyspark/sql/connect/expressions.py (hunk 4 above)
Comment (zhengruifeng) [reply python/pyspark/sql/connect/expressions.py:448]: oh, `__repr__` is still needed for now.

it's not needed in plan, but we have not handle it for expressions for now.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
I am making a followup to enable the related doctests. I will address if there are some more comments in this PR.

**Pull request closed without merging**